include/swift/ABI/MetadataValues.h

```diff
@@ -1237,18 +1253,74 @@ class TypeContextDescriptorFlags : public FlagSet<uint16_t> {
   explicit TypeContextDescriptorFlags(uint16_t bits) : FlagSet(bits) {}
   constexpr TypeContextDescriptorFlags() {}
 
-  FLAGSET_DEFINE_FLAG_ACCESSORS(IsCTag, isCTag, setIsCTag)
-  FLAGSET_DEFINE_FLAG_ACCESSORS(IsCTypedef, isCTypedef, setIsCTypedef)
   FLAGSET_DEFINE_FLAG_ACCESSORS(IsReflectable, isReflectable, setIsReflectable)
 
+  enum MetadataInitializationKind {
+    /// There are either no special rules for initializing the metadata
+    /// or the metadata is generic.  (Genericity is set in the
+    /// non-kind-specific descriptor flags.)
+    NoMetadataInitialization = 0,
+
+    /// The type requires non-trivial singleton initialization using the
+    /// "in-place" code pattern.
+    InPlaceMetadataInitialization = 1,
+
+    // We only have two bits here, so if you add a third special kind,
+    // include more flag bits in its out-of-line storage.
+  };
+
+  FLAGSET_DEFINE_FIELD_ACCESSORS(MetadataInitialization,
+                                 MetadataInitialization_width,
+                                 MetadataInitializationKind,
+                                 getMetadataInitialization,
+                                 setMetadataInitialization)
+
+  bool hasInPlaceMetadataInitialization() const {
+    return getMetadataInitialization() == InPlaceMetadataInitialization;
+  }
+
+  enum ImportNamespaceKind {
+    /// The type comes the default namespace for its language.
+    DefaultNamespace = 0,
+
+    // The behavior for C imported types is complicated in ways that don't
+    // entirely make sense according to the design laid out in the comment
+    // on the ImportNamespace field.  The rules are basically:
+    //   - Classes are assumed to come from Objective-C by default.
+    //     ObjC classes are in the ordinary namespace in C.
+    //   - Protocols are assumed to come from Objective-C by default.
+    //     ObjC protocols are in their own namespace in C.
+    //   - Structs and enums seem to always get either CTag or CTypedef.
+    //     It would probably make more sense to assume they come from the
+    //     tag namespace in C and then just use CTypedef as an override.
```

Comment:
(This part is still fine, though. We would never map a class or protocol to a struct.)